web_src/js/features/user-auth-webauthn.js

```diff
@@ -0,0 +1,243 @@
+        publicKey: makeAssertionOptions.publicKey
+      })
+        .then((credential) => {
+          verifyAssertion(credential);
+        }).catch((err) => {
+          webAuthnError(err.message);
+        });
+    });
+}
+
+function verifyAssertion(assertedCredential) {
+  // Move data into Arrays incase it is super long
+  const authData = new Uint8Array(assertedCredential.response.authenticatorData);
+  const clientDataJSON = new Uint8Array(assertedCredential.response.clientDataJSON);
+  const rawId = new Uint8Array(assertedCredential.rawId);
+  const sig = new Uint8Array(assertedCredential.response.signature);
+  const userHandle = new Uint8Array(assertedCredential.response.userHandle);
+  $.ajax({
+    url: `${appSubUrl}/user/webauthn/assertion`,
+    type: 'POST',
+    data: JSON.stringify({
+      id: assertedCredential.id,
+      rawId: bufferEncode(rawId),
+      type: assertedCredential.type,
+      response: {
+        authenticatorData: bufferEncode(authData),
+        clientDataJSON: bufferEncode(clientDataJSON),
+        signature: bufferEncode(sig),
+        userHandle: bufferEncode(userHandle),
+      },
+    }),
+    contentType: 'application/json; charset=utf-8',
+    dataType: 'json',
+    success: (resp) => {
+      if (resp && resp['redirect']) {
+        window.location.href = resp['redirect'];
+      } else {
+        window.location.href = '/';
+      }
+    },
+    error: (ret) => {
+      console.error('3-----', ret);
+    }
+  });
+}
+
+const lookup = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/';
+
+function encode (num) {
+  return lookup.charAt(num);
+}
+
+function tripletToBase64 (num) {
+  return encode(num >> 18 & 0x3F) + encode(num >> 12 & 0x3F) + encode(num >> 6 & 0x3F) + encode(num & 0x3F);
+}
+
+function uint8ToBase64 (uint8) {
+  const extraBytes = uint8.length % 3; // if we have 1 byte left, pad 2 bytes
+  let output = '';
+  let temp = 0;
+
+  // go through the array every three bytes, we'll deal with trailing stuff later
+  for (let i = 0, length = uint8.length - extraBytes; i < length; i += 3) {
+    temp = (uint8[i] << 16) + (uint8[i + 1] << 8) + (uint8[i + 2]);
+    output += tripletToBase64(temp);
+  }
+
+  // pad the end with zeros, but make sure to not forget the extra bytes
+  switch (extraBytes) {
+    case 1:
+      temp = uint8[uint8.length - 1];
+      output += encode(temp >> 2);
+      output += encode((temp << 4) & 0x3F);
+      output += '==';
+      break;
+    case 2:
+      temp = (uint8[uint8.length - 2] << 8) + (uint8[uint8.length - 1]);
+      output += encode(temp >> 10);
+      output += encode((temp >> 4) & 0x3F);
+      output += encode((temp << 2) & 0x3F);
+      output += '=';
+      break;
+    default:
+      break;
+  }
+
+  return output;
+}
+
+// Encode an ArrayBuffer into a base64 string.
+function bufferEncode(value) {
+  return uint8ToBase64(value)
+    .replace(/\+/g, '-')
+    .replace(/\//g, '_')
+    .replace(/=/g, '');
+}
```

Comment:
Maybe we can use a dependency like https://www.npmjs.com/package/base64-arraybuffer for this.

Reply:
Actually I guess it may just be fine to keep the code. It's small and every dependency poses a certain risk.